include/bluetooth/bluetooth.h

```diff
@@ -504,15 +523,9 @@ struct bt_le_adv_param {
  *                    address of peer for directed advertising.
  */
 #define BT_LE_ADV_PARAM(_options, _int_min, _int_max, _peer) \
-		((struct bt_le_adv_param[]) { { \
-			.id = BT_ID_DEFAULT, \
-			.sid = 0, \
-			.secondary_max_skip = 0, \
-			.options = (_options), \
-			.interval_min = (_int_min), \
-			.interval_max = (_int_max), \
-			.peer = (_peer), \
-		 } })
+	((struct bt_le_adv_param[]) { \
+		BT_LE_ADV_PARAM_INIT(_options, _int_min, _int_max, _peer) \
```

Comment:
This doesn't work; the cast is still of a temporary.  Using a version of `samples/bluetooth/peripheral` converted to C++:
```
/mnt/nordic/zp/zephyr/include/bluetooth/bluetooth.h:526:30: error: taking address of temporary array
  526 |  ((struct bt_le_adv_param[]) { \
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
  527 |   BT_LE_ADV_PARAM_INIT(_options, _int_min, _int_max, _peer) \
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  528 |   })
      |   ~~                          
/mnt/nordic/zp/zephyr/include/bluetooth/bluetooth.h:539:29: note: in expansion of macro 'BT_LE_ADV_PARAM'
  539 | #define BT_LE_ADV_CONN_NAME BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE | \
      |                             ^~~~~~~~~~~~~~~
../src/main.cxx:260:24: note: in expansion of macro 'BT_LE_ADV_CONN_NAME'
  260 |  err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
      |                        ^~~~~~~~~~~~~~~~~~~
```

The diagnostic goes away if the cast type is changed to `const struct bt_le_adv_param[]`

Reply:
@pabigot Thanks. I didn't add the const-qualifier, because the recommendation stated that was best to avoid these constructs altogether for C++. Which is why I added _INIT macros and removed the use in the static inline functions.
C++ code can then include the header, but it must avoid these specific macros.

If I added const to all the places that used this I get a ripple effect of const qualifier being discarded.

If you change your C++ peripheral like this:
```
struct bt_le_adv_param param = BT_LE_ADV_PARAM_INIT(BT_LE_ADV_OPT_CONNECTABLE |
						    BT_LE_ADV_OPT_USE_NAME,
						    BT_GAP_ADV_FAST_INT_MIN_2,
						    BT_GAP_ADV_FAST_INT_MAX_2,
						    NULL);
err = bt_le_adv_start(&param, ad, ARRAY_SIZE(ad), NULL, 0);
```
You should be good.

Reply:
OK, I accept that's a workaround so I'll remove my NAK.  It'd be nice to have uses of those macros in C++ code produce a diagnostic, but that's not in scope here.

The Zephyr Bluetooth examples are awash in uses of the original (non-init) macro as an argument to functions.  A goal of the original solution was to keep the recommended calling code the same for both C and C++ code.  Clearly that's not feasible.

We should probably add or convert one of the bluetooth samples to C++ to show how it's supposed to be done.

FWIW include/bluetooth/gatt.h at BT_GATT_CHARACTERISTIC also has an explicit cast of a temporary compound literal being cast to an array. That may not be as easy to fix, but also need not be part of this PR.